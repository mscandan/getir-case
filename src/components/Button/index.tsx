import React from 'react';
import styled from 'styled-components';

const StyledButton = styled.button`
  display: flex;
  align-items: center;
  justify-content: center;
  height: 22px;
  cursor: pointer;
  border-radius: 2px;
  font-weight: 600;
  font-size: 12px;
  line-height: 20px;
  background-color: ${({ theme }) => theme.COMPONENT_BG.BASE};
  color: ${({ theme }) => theme.TEXT_LIGHT.HIGHLIGHT};

  &:hover {
    cursor: pointer;
  }
`;

const StyledButtonContentContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
`;

export const Button: React.FC<React.ButtonHTMLAttributes<HTMLButtonElement> & React.PropsWithChildren> = React.memo(
  ({ children, onClick, type = 'button', ...rest }) => {
    return (
      <StyledButton type={type} onClick={onClick} {...rest} data-testid="button">
        <StyledButtonContentContainer>{children}</StyledButtonContentContainer>
      </StyledButton>
    );
  },
);
