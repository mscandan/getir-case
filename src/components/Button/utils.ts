import { DefaultTheme } from 'styled-components';
import { ButtonSize, ButtonType } from 'types';

export const getBorderStyling = (theme: DefaultTheme) => {
  return `2px solid ${theme.COMPONENT_BG.BASE}`;
};

export const getFontSize = (size: ButtonSize) => {
  switch (size) {
    case 'md':
      return 13;
    case 'lg':
      return 14;
    default:
      return 12;
  }
};

export const getBackgroundColor = (theme: DefaultTheme, buttonType: ButtonType) => {
  switch (buttonType) {
    case 'solid':
      return theme.COMPONENT_BG.BASE;
    default:
      return 'transparent';
  }
};

export const getSize = (size: ButtonSize) => {
  switch (size) {
    case 'lg':
      return '16px 24px';
    case 'md':
      return '16px 6px';
    default:
      return '6px 2px';
  }
};

export const getColor = (theme: DefaultTheme, buttonType: ButtonType) => {
  switch (buttonType) {
    case 'solid':
      return theme.TEXT_LIGHT.HIGHLIGHT;
    case 'outline':
      return theme.COMPONENT_BG.BASE;
    default:
      return theme.TEXT_DARK.BASE;
  }
};
