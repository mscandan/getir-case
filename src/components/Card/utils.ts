import { CardSizeType } from 'types';

export const getCardWidth = (size: CardSizeType) => {
  switch (size) {
    case 'sm':
      return '296px';
    case 'lg':
      return '608px';
    default:
      return '100%';
  }
};
