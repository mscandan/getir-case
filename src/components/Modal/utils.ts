export const isGivenParameterNumber = (parameter: number | string) => typeof parameter === 'number';
