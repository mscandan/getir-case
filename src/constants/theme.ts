import { DefaultTheme } from 'styled-components';

export const appTheme: DefaultTheme = {
  PAGE_BG: { BASE: '#E5E5E5' },
  COMPONENT_BG: { BASE: '#1EA4CE', HIGHLIGHT: '#F2F0FD', LOWLIGHT: '#147594' },
  TEXT_LIGHT: { BASE: '#A8A8A8', HIGHLIGHT: '#FFFFFF', LOWLIGHT: '#697488' },
  TEXT_DARK: { BASE: '#525252', HIGHLIGHT: '#E0E0E0', LOWLIGHT: '#191919' },
  BORDER: { BASE: '#DFDEE2', LOWLIGHT: '#F4F4F4' },
} as const;
