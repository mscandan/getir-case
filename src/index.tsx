import React from 'react';
import ReactDOM from 'react-dom/client';
import { Provider as ReduxProvider } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react';
import { store, persistor } from 'store';
import App from './App';

const root = ReactDOM.createRoot(document.getElementById('root') as HTMLElement);
root.render(
  <React.StrictMode>
    <ReduxProvider store={store}>
      <PersistGate persistor={persistor}>
        <App />
      </PersistGate>
    </ReduxProvider>
  </React.StrictMode>,
);
