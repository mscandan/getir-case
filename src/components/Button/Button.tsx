import React from 'react';
import styled from 'styled-components';
import { ButtonSize, ButtonType } from 'types';
import { getBackgroundColor, getBorderStyling, getColor, getFontSize, getSize } from './utils';

export interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  size: ButtonSize;
  buttonType?: ButtonType;
}

const StyledButton = styled.button<{ size: ButtonSize; buttonType: ButtonType }>`
  border: ${({ theme, buttonType }) => buttonType === 'outline' && getBorderStyling(theme)};
  font-size: ${({ size }) => getFontSize(size)}px;
  background-color: ${({ theme, buttonType }) => getBackgroundColor(theme, buttonType)};
  padding: ${({ size }) => getSize(size)};
  color: ${({ theme, buttonType }) => getColor(theme, buttonType)};
  border-radius: 2px;

  &:hover {
    cursor: pointer;
  }
`;

const StyledButtonContentContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
`;

export const Button: React.FC<ButtonProps & React.PropsWithChildren> = React.memo(
  ({ children, onClick, type = 'button', size, buttonType = 'solid', ...rest }) => {
    return (
      <StyledButton type={type} onClick={onClick} size={size} buttonType={buttonType} {...rest} data-testid="button">
        <StyledButtonContentContainer>{children}</StyledButtonContentContainer>
      </StyledButton>
    );
  },
);
