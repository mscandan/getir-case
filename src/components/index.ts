import { Button } from './Button';
import { Checkbox } from './Checkbox';
import { Radios } from './Radios';
import { Icon } from './Icon';
import { Card } from './Card';
import { ButtonGroup } from './ButtonGroup';
import { Input } from './Input';
import { Counter } from './Counter';
import { Dropdown } from './Dropdown';
import { LoadingSpinner } from './LoadingSpinner';
import { ProductItem } from './ProductItem';

export { Button, Checkbox, Radios, Icon, Card, ButtonGroup, Input, Counter, Dropdown, LoadingSpinner, ProductItem };
