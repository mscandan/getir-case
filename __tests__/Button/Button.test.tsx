import React from 'react';
import renderer from 'react-test-renderer';
import { fireEvent, render } from '@testing-library/react';
import { componentRenderer } from '../utils';
import { FlatButton, OutlineButton, SolidButton } from '../../src/components';
import { getBackgroundColor, getBorderStyling, getColor, getSize } from '../../src/components/Button/utils';
import { appTheme } from '../../src/constants/theme';

describe('Button Tests', () => {
  describe('Solid Button Tests', () => {
    it('should match snapshot with small solid button', () => {
      const comp = renderer.create(componentRenderer(<SolidButton size="sm">hello</SolidButton>));
      expect(comp).toMatchSnapshot();
    });

    it('should match snapshot with medium solid button', () => {
      const comp = renderer.create(componentRenderer(<SolidButton size="md">hello</SolidButton>));
      expect(comp).toMatchSnapshot();
    });

    it('should match snapshot with large solid button', () => {
      const comp = renderer.create(componentRenderer(<SolidButton size="lg">hello</SolidButton>));
      expect(comp).toMatchSnapshot();
    });

    it('should render the given child component in solid button', () => {
      const { getByText } = render(
        componentRenderer(
          <SolidButton size="sm">
            <div>hello</div>
          </SolidButton>,
        ),
      );
      expect(getByText('hello')).toBeInTheDocument();
    });

    it('should invoke the given function on click in solid button', () => {
      const onClick = jest.fn();

      const { getByTestId } = render(
        componentRenderer(
          <SolidButton size="sm" onClick={onClick}>
            <div>hello</div>
          </SolidButton>,
        ),
      );
      fireEvent.click(getByTestId('button'));
      expect(onClick).toBeCalledTimes(1);
    });
  });

  describe('Flat Button Tests', () => {
    it('should match snapshot with flat button', () => {
      const comp = renderer.create(componentRenderer(<FlatButton size="sm">hello</FlatButton>));
      expect(comp).toMatchSnapshot();
    });

    it('should match snapshot with medium flat button', () => {
      const comp = renderer.create(componentRenderer(<FlatButton size="md">hello</FlatButton>));
      expect(comp).toMatchSnapshot();
    });

    it('should match snapshot with large flat button', () => {
      const comp = renderer.create(componentRenderer(<FlatButton size="lg">hello</FlatButton>));
      expect(comp).toMatchSnapshot();
    });

    it('should render the given child component in flat button', () => {
      const { getByText } = render(
        componentRenderer(
          <FlatButton size="sm">
            <div>hello</div>
          </FlatButton>,
        ),
      );
      expect(getByText('hello')).toBeInTheDocument();
    });

    it('should invoke the given function on click in flat button', () => {
      const onClick = jest.fn();

      const { getByTestId } = render(
        componentRenderer(
          <FlatButton size="sm" onClick={onClick}>
            <div>hello</div>
          </FlatButton>,
        ),
      );
      fireEvent.click(getByTestId('button'));
      expect(onClick).toBeCalledTimes(1);
    });
  });

  describe('Outline Button Tests', () => {
    it('should match snapshot with outline button', () => {
      const comp = renderer.create(componentRenderer(<OutlineButton size="sm">hello</OutlineButton>));
      expect(comp).toMatchSnapshot();
    });

    it('should match snapshot with medium outline button', () => {
      const comp = renderer.create(componentRenderer(<FlatButton size="md">hello</FlatButton>));
      expect(comp).toMatchSnapshot();
    });

    it('should match snapshot with large outline button', () => {
      const comp = renderer.create(componentRenderer(<FlatButton size="lg">hello</FlatButton>));
      expect(comp).toMatchSnapshot();
    });

    it('should render the given child component in outline button', () => {
      const { getByText } = render(
        componentRenderer(
          <OutlineButton size="sm">
            <div>hello</div>
          </OutlineButton>,
        ),
      );
      expect(getByText('hello')).toBeInTheDocument();
    });

    it('should invoke the given function on click in outline button', () => {
      const onClick = jest.fn();

      const { getByTestId } = render(
        componentRenderer(
          <OutlineButton size="sm" onClick={onClick}>
            <div>hello</div>
          </OutlineButton>,
        ),
      );
      fireEvent.click(getByTestId('button'));
      expect(onClick).toBeCalledTimes(1);
    });
  });

  describe('Utility Function Tests', () => {
    it('should return proper border styling', () => {
      const style = getBorderStyling(appTheme);
      expect(style).toEqual('2px solid #1EA4CE');
    });

    it('should return proper background styling', () => {
      let style = getBackgroundColor(appTheme, 'solid');
      expect(style).toEqual('#1EA4CE');
      style = getBackgroundColor(appTheme, 'flat');
      expect(style).toEqual('transparent');
      style = getBackgroundColor(appTheme, 'outline');
      expect(style).toEqual('transparent');
    });

    it('should return proper padding sizes', () => {
      let style = getSize('sm');
      expect(style).toEqual('6px 2px');
      style = getSize('md');
      expect(style).toEqual('16px 6px');
      style = getSize('lg');
      expect(style).toEqual('16px 24px');
    });

    it('should return proper color styling', () => {
      let style = getColor(appTheme, 'flat');
      expect(style).toEqual('#525252');
      style = getColor(appTheme, 'solid');
      expect(style).toEqual('#FFFFFF');
      style = getColor(appTheme, 'outline');
      expect(style).toEqual('#1EA4CE');
    });
  });
});
