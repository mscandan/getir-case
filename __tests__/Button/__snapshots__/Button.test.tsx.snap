// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Tests should match snapshot 1`] = `
<button
  className="sc-bczRLJ bDyXjI"
  data-testid="button"
  type="button"
>
  Hello
</button>
`;
